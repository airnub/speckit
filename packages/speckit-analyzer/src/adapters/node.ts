import { promises as fs } from "node:fs";
import path from "node:path";

import type { FailureRule, RawLogSource } from "../types.js";
import { parseFailureRules } from "../rules.js";

export interface FileLogSourceOptions {
  id?: string;
  encoding?: BufferEncoding;
  format?: RawLogSource["format"];
}

export async function createFileLogSource(
  filePath: string,
  options: FileLogSourceOptions = {}
): Promise<RawLogSource> {
  const raw = await fs.readFile(filePath, { encoding: options.encoding ?? "utf8" });
  return {
    id: options.id ?? filePath,
    content: raw,
    format: options.format,
  };
}

export async function loadLogSourcesFromFiles(
  paths: string[],
  options: FileLogSourceOptions = {}
): Promise<RawLogSource[]> {
  return Promise.all(paths.map((filePath) => createFileLogSource(filePath, { ...options, id: options.id ?? filePath })));
}

export async function loadFailureRulesFromFs(
  rootDir: string,
  artifactDir?: string
): Promise<FailureRule[]> {
  const baseDir = artifactDir ?? path.join(rootDir, ".speckit");
  const rulesPath = path.join(baseDir, "failure-rules.yaml");
  try {
    const raw = await fs.readFile(rulesPath, "utf8");
    return parseFailureRules(raw);
  } catch (error) {
    console.warn(`[analyzer:node] Unable to load failure rules from ${rulesPath}: ${(error as Error).message}`);
    return [];
  }
}
