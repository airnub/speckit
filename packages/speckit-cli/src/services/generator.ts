import { createHash } from "node:crypto";
import path from "node:path";
import fs from "fs-extra";
import nunjucks from "nunjucks";
import matter from "gray-matter";
import {
  loadSpecYaml,
  hashSpecYaml,
} from "./spec.js";
import {
  loadCatalogLock,
  loadBundle,
  assertSpeckitCompatibility,
  assertSpecCompatibility,
} from "./catalog.js";
import { getSpeckitVersion } from "./version.js";
import { appendManifestRun } from "./manifest.js";

type Provenance = {
  tool: "speckit";
  tool_version: string;
  tool_commit: string;
  template: { id: string; version: string; sha: string };
  spec: { version: string; digest: string };
  generated_at: string;
};

type PreparedOutput = {
  relPath: string;
  absPath: string;
  content: string;
  provenance: Provenance;
  digest: string;
  changed: boolean;
};

type PrepareOutputOptions = {
  outputId?: string;
  specMeta?: unknown;
};

export type GenerateOptions = {
  repoRoot?: string;
  write: boolean;
  stdout?: NodeJS.WritableStream;
};

export type GenerateResult = {
  outputs: { path: string; changed: boolean }[];
};

export async function generateDocs(options: GenerateOptions): Promise<GenerateResult> {
  const repoRoot = options.repoRoot ?? process.cwd();
  const { data: specData } = await loadSpecYaml(repoRoot);
  const spec = specData?.spec;
  if (!spec) {
    throw new Error("spec.yaml missing root 'spec' object");
  }
  const specMeta = spec.meta ?? {};
  if (!specMeta || typeof specMeta.version !== "string" || !specMeta.version.trim()) {
    throw new Error("spec.meta.version is required");
  }
  const specVersion = specMeta.version.trim();
  const specDigest = await hashSpecYaml(repoRoot);

  const speckitInfo = await getSpeckitVersion(repoRoot);
  const catalogEntries = await loadCatalogLock(repoRoot);
  if (!catalogEntries.length) {
    throw new Error(".speckit/catalog.lock has no bundles");
  }

  const results: { path: string; changed: boolean }[] = [];

  for (const entry of catalogEntries) {
    const bundle = await loadBundle(repoRoot, entry);
    assertSpeckitCompatibility(speckitInfo.version, entry, bundle);
    assertSpecCompatibility(specVersion, bundle);

    const env = nunjucks.configure(bundle.dir, {
      autoescape: false,
      throwOnUndefined: true,
      noCache: true,
    });

    const context = createRenderingContext(specData, spec);
    const runTimestamp = new Date().toISOString();

    const preparedOutputs: PreparedOutput[] = [];

    for (const output of bundle.outputs) {
      const templatePath = path.join(bundle.dir, output.from);
      const templateSource = await fs.readFile(templatePath, "utf8");
      registerLiteralPlaceholders(env, templateSource);
      const rendered = env.render(output.from, context);
      const targetRel = env.renderString(output.to, context).trim();
      if (!targetRel) {
        throw new Error(`Bundle '${bundle.id}' produced empty target path for output '${output.id}'`);
      }
      const targetPath = path.join(repoRoot, targetRel);
      const existing = await readIfExists(targetPath);

      const baseProvenance: Provenance = {
        tool: "speckit",
        tool_version: speckitInfo.version,
        tool_commit: speckitInfo.commit,
        template: { id: bundle.id, version: bundle.version, sha: entry.sha },
        spec: { version: specVersion, digest: specDigest },
        generated_at: runTimestamp,
      };

      const prepared = prepareOutput(rendered, targetRel, baseProvenance, existing, {
        outputId: output.id,
        specMeta,
      });

      const digest = hashContent(prepared.content);
      const changed = existing === null || normalise(existing) !== normalise(prepared.content);

      preparedOutputs.push({
        relPath: targetRel,
        absPath: targetPath,
        content: prepared.content,
        provenance: prepared.provenance,
        digest,
        changed,
      });

      results.push({ path: targetRel, changed });
    }

    const changedOutputs = preparedOutputs.filter(p => p.changed);

    if (options.write) {
      for (const output of changedOutputs) {
        await fs.ensureDir(path.dirname(output.absPath));
        await fs.writeFile(output.absPath, output.content, "utf8");
        options.stdout?.write(`Updated ${output.relPath}\n`);
      }

      if (changedOutputs.length > 0) {
        const first = changedOutputs[0];
        await appendManifestRun(repoRoot, speckitInfo, {
          at: first.provenance.generated_at,
          spec: { version: specVersion, digest: specDigest },
          template: {
            id: bundle.id,
            version: bundle.version,
            sha: entry.sha,
          },
          outputs: changedOutputs.map(o => ({ path: o.relPath, digest: o.digest })),
        });
      }
    }
  }

  return { outputs: results };
}

function normalise(value: string): string {
  return value.replace(/\r\n/g, "\n");
}

async function readIfExists(filePath: string): Promise<string | null> {
  if (!(await fs.pathExists(filePath))) return null;
  return await fs.readFile(filePath, "utf8");
}

function hashContent(content: string): string {
  const digest = createHash("sha256").update(content, "utf8").digest("hex");
  return `sha256:${digest}`;
}

function prepareOutput(
  rendered: string,
  targetRel: string,
  provenance: Provenance,
  existing: string | null,
  options: PrepareOutputOptions = {}
): { content: string; provenance: Provenance } {
  const ext = path.extname(targetRel).toLowerCase();
  if (ext === ".md" || ext === ".mdx") {
    return prepareMarkdown(rendered, provenance, existing, options);
  }
  return prepareWithComment(rendered, provenance, existing, ext);
}

function prepareMarkdown(
  rendered: string,
  provenance: Provenance,
  existing: string | null,
  options: PrepareOutputOptions = {}
): { content: string; provenance: Provenance } {
  const parsed = matter(normalise(rendered));
  const body = parsed.content;
  let finalProv = { ...provenance };

  if (existing) {
    const existingParsed = matter(normalise(existing));
    const existingProv = existingParsed.data?.speckit_provenance;
    if (existingProv && typeof existingProv === "object") {
      const { generated_at: prevGeneratedAt, ...restExisting } = existingProv as Record<string, unknown>;
      const { generated_at: newGeneratedAt, ...restNew } = provenance;
      const bodySame = normalise(existingParsed.content) === normalise(body);
      if (bodySame && deepEqual(restExisting, restNew) && typeof prevGeneratedAt === "string" && prevGeneratedAt.trim()) {
        finalProv = { ...provenance, generated_at: prevGeneratedAt.trim() };
      }
    }
  }

  const withWhyItMatters = maybeInjectWhyItMatters(body, options);
  const data = { ...parsed.data, speckit_provenance: finalProv };
  const stringified = matter.stringify(withWhyItMatters, data).replace(/\r\n/g, "\n");
  return { content: ensureTrailingNewline(stringified), provenance: finalProv };
}

function prepareWithComment(
  rendered: string,
  provenance: Provenance,
  existing: string | null,
  ext: string
): { content: string; provenance: Provenance } {
  const comment = buildComment(provenance, ext);
  const body = normalise(rendered).replace(/^\uFEFF/, "").trimStart();
  const candidate = ensureTrailingNewline(`${comment}\n${body}`);
  if (existing && normalise(existing) === normalise(candidate)) {
    return { content: ensureTrailingNewline(existing), provenance: provenance };
  }
  return { content: candidate, provenance };
}

function ensureTrailingNewline(value: string): string {
  return value.endsWith("\n") ? value : `${value}\n`;
}

function maybeInjectWhyItMatters(
  body: string,
  options: PrepareOutputOptions
): string {
  const whyItMatters = extractWhyItMatters(options);
  if (!whyItMatters || hasWhyItMattersSection(body)) {
    return body;
  }
  const trimmed = body.replace(/\s+$/, "");
  const section = formatWhyItMattersSection(whyItMatters);
  if (!trimmed) {
    return `${section}\n`;
  }
  return `${trimmed}\n\n${section}\n`;
}

function extractWhyItMatters(options: PrepareOutputOptions): string[] | undefined {
  if (!options || options.outputId !== "spec") {
    return undefined;
  }
  const meta = options.specMeta;
  if (!meta || typeof meta !== "object") {
    return undefined;
  }
  const candidate =
    (meta as Record<string, unknown>)["why_it_matters"] ??
    (meta as Record<string, unknown>)["whyItMatters"];
  if (!Array.isArray(candidate)) {
    return undefined;
  }
  const cleaned = candidate
    .filter((item): item is string => typeof item === "string")
    .map(item => item.trim())
    .filter(item => item.length > 0);
  return cleaned.length ? cleaned : undefined;
}

function hasWhyItMattersSection(body: string): boolean {
  return /(^|\n)##\s+Why it matters\b/i.test(body);
}

function formatWhyItMattersSection(points: string[]): string {
  const bullets = points.map(point => `- ${point}`).join("\n");
  return `## Why it matters\n\n${bullets}`;
}

function buildComment(provenance: Provenance, ext: string): string {
  const message = `Generated by Speckit ${provenance.tool_version} (${provenance.tool_commit}) using template ${provenance.template.id}@${provenance.template.sha} for spec ${provenance.spec.version} (${provenance.spec.digest}).`;
  switch (ext) {
    case ".yml":
    case ".yaml":
    case ".sh":
    case ".py":
      return `# ${message}`;
    case ".ts":
    case ".tsx":
    case ".js":
    case ".jsx":
    case ".json":
      return `/* ${message} */`;
    default:
      return `<!-- ${message} -->`;
  }
}

function createRenderingContext(specData: any, spec: any): any {
  const base = { ...specData, spec };
  const placeholderPattern = /^[A-Z0-9_]+$/;
  return new Proxy(base, {
    get(target, prop, receiver) {
      if (typeof prop === "string") {
        if (Reflect.has(target, prop)) {
          return Reflect.get(target, prop, receiver);
        }
        if (placeholderPattern.test(prop)) {
          return `{{${prop}}}`;
        }
      }
      return Reflect.get(target, prop, receiver);
    },
    has(target, prop) {
      if (Reflect.has(target, prop)) return true;
      return typeof prop === "string" && placeholderPattern.test(prop);
    },
  });
}

function registerLiteralPlaceholders(env: nunjucks.Environment, source: string) {
  const regex = /\{\{\s*([A-Z0-9_]+)\s*\}\}/g;
  let match: RegExpExecArray | null;
  while ((match = regex.exec(source))) {
    const name = match[1];
    env.addGlobal(name, `{{${name}}}`);
  }
}

function deepEqual(a: unknown, b: unknown): boolean {
  return stableStringify(a) === stableStringify(b);
}

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(v => stableStringify(v)).join(",")}]`;
  }
  const entries = Object.entries(value as Record<string, unknown>)
    .map(([key, val]) => [key, stableStringify(val)] as const)
    .sort(([a], [b]) => a.localeCompare(b));
  return `{${entries.map(([key, val]) => `${JSON.stringify(key)}:${val}`).join(",")}}`;
}
