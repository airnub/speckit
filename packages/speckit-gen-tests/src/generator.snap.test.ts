import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import fs from "fs-extra";
import { afterAll, describe, expect, it, vi } from "vitest";

import type { SpecModel } from "../../speckit-core/src/model/SpecModel.js";
import type { BundleDefinition, CatalogLockEntry } from "../../speckit-cli/src/services/catalog.js";
import { generateDocs } from "../../speckit-cli/src/services/generator.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const defaultBundleDir = path.join(__dirname, "fixtures", "test-bundle");

const baseModel: SpecModel = {
  version: "2024.09",
  meta: { title: "SpecKit Demo", mode: "Classic" },
  requirements: [
    { id: "REQ-001", title: "Render overview" },
    { id: "REQ-002", title: "Preserve placeholders" },
  ],
};

const mockDialect = { id: "speckit.v1", version: "1.1.0" };

let activeModel: SpecModel = deepClone(baseModel);
let activeBundle: BundleDefinition = createBundleDefinition({
  id: "test-bundle",
  dir: defaultBundleDir,
  outputs: [
    { id: "overview", from: "overview.md.njk", to: "docs/overview.md" },
    { id: "config", from: "config.yaml.njk", to: "config/app.yaml" },
    { id: "metadata", from: "metadata.ts.njk", to: "src/metadata.ts" },
  ],
});
let activeCatalogEntry: CatalogLockEntry = createCatalogEntry(activeBundle);

vi.mock("../../speckit-cli/src/services/spec.js", () => ({
  loadSpecModel: vi.fn(async () => ({
    model: mockModel,
    dialect: mockDialect,
    data: { engine: { mode: "classic" } },
  })),
  hashSpecYaml: vi.fn(async () => "sha256:mock-spec-digest"),
}));

vi.mock("../../speckit-cli/src/services/catalog.js", () => ({
  loadCatalogLock: vi.fn(async () => [activeCatalogEntry]),
  loadBundle: vi.fn(async () => activeBundle),
  assertSpeckitCompatibility: vi.fn(),
  assertSpecCompatibility: vi.fn(),
  assertDialectCompatibility: vi.fn(),
}));

vi.mock("../../speckit-cli/src/services/version.js", async () => {
  const actual = await vi.importActual<typeof import("../../speckit-cli/src/services/version.js")>(
    "../../speckit-cli/src/services/version.js"
  );
  return {
    ...actual,
    getSpeckitVersion: vi.fn(async () => ({ version: "0.1.0", commit: "abc1234" })),
  };
});

vi.mock("../../speckit-cli/src/services/manifest.js", () => ({
  appendManifestRun: vi.fn(),
  updateManifestSpeckit: vi.fn(),
}));

const tempRepos: string[] = [];

async function createRepoRoot(): Promise<string> {
  const dir = await fs.mkdtemp(path.join(os.tmpdir(), "speckit-gen-tests-"));
  tempRepos.push(dir);
  return dir;
}

afterAll(async () => {
  for (const dir of tempRepos) {
    await fs.remove(dir);
  }
});

describe("generator snapshots", () => {
  beforeEach(() => {
    resetActiveFixtures();
  });

  it("renders markdown output with provenance snapshot", async () => {
    vi.useFakeTimers();
    try {
      vi.setSystemTime(new Date("2024-01-02T03:04:05Z"));
      const repoRoot = await createRepoRoot();
      await generateDocs({ repoRoot, write: true });
      const markdown = await fs.readFile(path.join(repoRoot, "docs", "overview.md"), "utf8");
      expect(markdown).toMatchSnapshot();
    } finally {
      vi.useRealTimers();
    }
  });

  it("renders yaml output with language appropriate header", async () => {
    vi.useFakeTimers();
    try {
      vi.setSystemTime(new Date("2024-01-02T03:04:05Z"));
      const repoRoot = await createRepoRoot();
      await generateDocs({ repoRoot, write: true });
      const yaml = await fs.readFile(path.join(repoRoot, "config", "app.yaml"), "utf8");
      expect(yaml).toMatchSnapshot();
    } finally {
      vi.useRealTimers();
    }
  });

  it("renders typescript metadata with provenance comment", async () => {
    vi.useFakeTimers();
    try {
      vi.setSystemTime(new Date("2024-01-02T03:04:05Z"));
      const repoRoot = await createRepoRoot();
      await generateDocs({ repoRoot, write: true });
      const tsContent = await fs.readFile(path.join(repoRoot, "src", "metadata.ts"), "utf8");
      expect(tsContent).toMatchSnapshot();
    } finally {
      vi.useRealTimers();
    }
  });

  it("retains provenance timestamp when rerun without changes", async () => {
    vi.useFakeTimers();
    try {
      const repoRoot = await createRepoRoot();
      vi.setSystemTime(new Date("2024-01-02T03:04:05Z"));
      const first = await generateDocs({ repoRoot, write: true });
      expect(first.outputs.every(output => output.changed)).toBe(true);
      const initial = await fs.readFile(path.join(repoRoot, "docs", "overview.md"), "utf8");
      expect(initial).toContain("generated_at: '2024-01-02T03:04:05.000Z'");

      vi.setSystemTime(new Date("2024-05-06T07:08:09Z"));
      const second = await generateDocs({ repoRoot, write: true });
      expect(second.outputs.every(output => output.changed === false)).toBe(true);
      const after = await fs.readFile(path.join(repoRoot, "docs", "overview.md"), "utf8");
      expect(after).toBe(initial);
    } finally {
      vi.useRealTimers();
    }
  });
});

describe("mode generator snapshots", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
    resetActiveFixtures();
  });

  const cases = [
    {
      label: "classic",
      dir: "mode-classic",
      bundleId: "mode-classic",
      outputPath: ["docs", "mode-classic.md"],
      meta: { title: "Mode Classic", mode: "Classic" },
    },
    {
      label: "secure",
      dir: "mode-secure",
      bundleId: "mode-secure",
      outputPath: ["docs", "mode-secure.md"],
      meta: { title: "Mode Secure", mode: "Secure" },
    },
  ] as const;

  it.each(cases)("renders $label mode snapshot", async fixture => {
    const repoRoot = await createRepoRoot();
    activeModel = {
      ...deepClone(baseModel),
      meta: { ...deepClone(baseModel.meta ?? {}), ...fixture.meta },
    };
    const bundleDir = path.join(__dirname, "fixtures", fixture.dir);
    activeBundle = createBundleDefinition({
      id: fixture.bundleId,
      dir: bundleDir,
      outputs: [
        { id: "mode", from: "mode.md.njk", to: `docs/${fixture.bundleId}.md` },
      ],
    });
    activeCatalogEntry = createCatalogEntry(activeBundle);

    vi.setSystemTime(new Date("2024-01-02T03:04:05Z"));
    await generateDocs({ repoRoot, write: true });
    const markdown = await fs.readFile(path.join(repoRoot, ...fixture.outputPath), "utf8");
    expect(markdown).toMatchSnapshot();
  });
});

type BundleShape = {
  id: string;
  dir: string;
  outputs: { id: string; from: string; to: string }[];
};

function createBundleDefinition(shape: BundleShape): BundleDefinition {
  return {
    id: shape.id,
    kind: "specs",
    version: "0.0.1",
    engine: "nunjucks",
    requires_speckit: ">=0.0.0",
    requires_dialect: { id: "speckit.v1", range: ">=1.0.0 <2.0.0" },
    outputs: shape.outputs.map(output => ({ ...output })),
    dir: shape.dir,
  };
}

function createCatalogEntry(bundleDef: BundleDefinition): CatalogLockEntry {
  return {
    id: bundleDef.id,
    sha: `${bundleDef.id}-sha`,
    version: bundleDef.version,
    requires_speckit: bundleDef.requires_speckit,
    requires_dialect: { ...bundleDef.requires_dialect },
    synced_with: { version: bundleDef.version, commit: `${bundleDef.id}-commit` },
  };
}

function resetActiveFixtures() {
  activeModel = deepClone(baseModel);
  activeBundle = createBundleDefinition({
    id: "test-bundle",
    dir: defaultBundleDir,
    outputs: [
      { id: "overview", from: "overview.md.njk", to: "docs/overview.md" },
      { id: "config", from: "config.yaml.njk", to: "config/app.yaml" },
      { id: "metadata", from: "metadata.ts.njk", to: "src/metadata.ts" },
    ],
  });
  activeCatalogEntry = createCatalogEntry(activeBundle);
}

function deepClone<T>(value: T): T {
  return JSON.parse(JSON.stringify(value));
}
