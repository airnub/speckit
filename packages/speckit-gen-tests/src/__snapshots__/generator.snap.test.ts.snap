// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generator snapshots > renders markdown output with provenance snapshot 1`] = `
"---
title: SpecKit Demo
release: 2024.09
speckit_provenance:
  tool: speckit
  tool_version: 0.1.0
  tool_commit: abc1234
  dialect:
    id: speckit.v1
    version: 1.1.0
  template:
    id: test-bundle
    version: 0.0.1
    sha: abc1234
  spec:
    version: '2024.09'
    digest: 'sha256:mock-spec-digest'
  generated_at: '2024-01-02T03:04:05.000Z'
---

# SpecKit Demo overview

Welcome to version 2024.09 of SpecKit Demo.

## Requirements
- REQ-001 — Render overview
- REQ-002 — Preserve placeholders



Status token: {{RELEASE_STATUS}}
"
`;

exports[`generator snapshots > renders typescript metadata with provenance comment 1`] = `
"/* Generated by Speckit 0.1.0 (abc1234) using template test-bundle@abc1234 for spec 2024.09 (sha256:mock-spec-digest) dialect speckit.v1@1.1.0. */
export const metadata = {
  name: "SpecKit Demo",
  version: "2024.09",
  requirements: 2,
};
"
`;

exports[`generator snapshots > renders yaml output with language appropriate header 1`] = `
"# Generated by Speckit 0.1.0 (abc1234) using template test-bundle@abc1234 for spec 2024.09 (sha256:mock-spec-digest) dialect speckit.v1@1.1.0.
name: speckit-demo-config
version: 2024.09
dialect: speckit.v1@1.1.0
requirements:
- id: REQ-001
    title: Render overview
- id: REQ-002
    title: Preserve placeholders

"
`;
